server/proxy/bindings.go

```diff
@@ -86,122 +91,156 @@ func (p *Proxy) GetAppBindings(in Incoming, cc apps.Context, app apps.App) []app
-				allowed = true
-				break
-			}
+		fql := locPrefix.Sub(clean.Location)
+		if usedLocations[clean.Location] {
+			problems = append(problems,
+				errors.Errorf("ignored diplicate command binding for location %q", clean.Location))
+			continue
 		}
-		if !allowed {
-			log.Debugw("location is not granted to app", "location", fql)
+		if fql.In(apps.LocationCommand) && usedCommandLabels[clean.Label] {
+			problems = append(problems,
+				errors.Errorf("ignored diplicate command binding for label %q (location %q)", clean.Label, clean.Location))
 			continue
 		}
 
-		if fql.In(apps.LocationCommand) {
-			label := b.Label
-			if label == "" {
-				label = string(b.Location)
-			}
+		out = append(out, *clean)
+	}
 
-			if strings.ContainsAny(label, " \t") {
-				log.Debugw("Binding validation error: Command label has multiple words", "location", b.Location)
-				continue
-			}
-		}
+	return out, problems
+}
 
-		if fql.IsTop() {
-			if locationsUsed[b.Location] {
-				continue
-			}
-			locationsUsed[b.Location] = true
-		} else {
-			if b.Location == "" || b.Label == "" {
-				continue
-			}
-			if locationsUsed[b.Location] || labelsUsed[b.Label] {
-				continue
-			}
+func cleanAppBinding(
+	app apps.App,
+	b apps.Binding,
+	locPrefix apps.Location,
+	userAgent string,
+	conf config.Config,
+) (*apps.Binding, []error) {
+	var problems []error
+
+	// Cleanup Location.
+	if b.Location == "" {
+		b.Location = apps.Location(app.Manifest.AppID)
+	}
+	if trimmed := apps.Location(strings.TrimSpace(string(b.Location))); trimmed != b.Location {
+		problems = append(problems, errors.Errorf("trimmed whitespace from location %s", trimmed))
+		b.Location = trimmed
+	}
 
-			locationsUsed[b.Location] = true
-			labelsUsed[b.Label] = true
-			b.AppID = app.Manifest.AppID
+	fql := locPrefix.Sub(b.Location)
+	allowed := false
+	for _, grantedLoc := range app.GrantedLocations {
+		// TODO Why `grantedLoc.In(fql)`?
+		if fql.In(grantedLoc) || grantedLoc.In(fql) {
+			allowed = true
+			break
 		}
+	}
+	if !allowed {
+		problems = append(problems, utils.NewForbiddenError("location %q is not granted", fql))
+		return nil, problems
+	}
 
-		if b.Icon != "" {
-			icon, err := normalizeStaticPath(conf, app.AppID, b.Icon)
-			if err != nil {
-				log.WithError(err).Debugw("Invalid icon path in binding",
-					"app_id", app.AppID,
-					"icon", b.Icon)
-				b.Icon = ""
-			} else {
-				b.Icon = icon
-			}
+	// Cleanup AppID.
+	if !fql.IsTop() {
+		b.AppID = app.AppID
+	}
+
+	// Cleanup (command) label.
+	if fql != apps.LocationCommand && fql.In(apps.LocationCommand) {
+		// A command binding must have a valid label. Default to Location if needed.
+		if b.Label == "" {
+			b.Label = string(b.Location)
+		}
+		if trimmed := strings.TrimSpace(b.Label); trimmed != b.Label {
+			problems = append(problems, errors.Errorf("trimmed whitespace from label %s", trimmed))
+			b.Label = trimmed
 		}
+		if strings.ContainsAny(b.Label, " \t") {
```

Comment:
Does this catch all cases? What about `foo bar`?

Reply:
@hanzei please elaborate? 1/5 We could do regexp `\s`, but seems an overkill, ` ` and `\t` seem good enough?

Reply:
I didn't saw the white space before `\t`. LGTM :+1: